src/passes/LocalSubtyping.cpp

```diff
@@ -0,0 +1,129 @@
+    FindAll<T> operations(func->body);
+    for (auto* operation : operations.list) {
+      ret[operation->index].push_back(operation);
+    }
+    return ret;
+  }
+
+  void doWalkFunction(Function* func) {
+    auto varBase = func->getVarIndexBase();
+    auto numLocals = func->getNumLocals();
+
+    auto setsForLocal = getLocalOperations<LocalSet>(func);
+    auto getsForLocal = getLocalOperations<LocalGet>(func);
+
+    // Keep iterating while we find things to change. There can be chains like
+    // X -> Y -> Z where one change enables more. Note that we are O(N^2) on
+    // that atm, but it is a rare pattern as general optimizations
+    // (SimplifyLocals and CoalesceLocals) break up such things; also, even if
+    // we tracked changes more carefully we'd have the case of nested tees
+    // where we could still be O(N^2), so we'd need something more complex here
+    // involving topological sorting. Leave that for if the need arises.
+
+    // TODO: handle cycles of X -> Y -> X etc.
+
+    bool more;
+
+    do {
+      more = false;
+
+      // First, refinalize which will recompute least upper bounds on ifs and
+      // blocks, etc., potentially finding a more specific type. Note that
+      // that utility does not tell us if it changed anything, so we depend on
+      // the next step for knowing if there is more work to do.
+      ReFinalize().walkFunctionInModule(func, getModule());
+
+      // Second, find vars whose actual applied values allow a more specific
+      // type.
+
+      for (Index i = varBase; i < numLocals; i++) {
+        // Find all the types assigned to the var, and compute the optimal LUB.
+        // Note that we do not need to take into account the initial value of
+        // zero or null that locals have: that value has the type of the local,
+        // which is a supertype of all the assigned values anyhow. It will never
+        // be able to tell us of a more specific subtype that is possible.
+        std::unordered_set<Type> types;
+        for (auto* set : setsForLocal[i]) {
+          types.insert(set->value->type);
+        }
+        if (types.empty()) {
+          continue;
+        }
+
+        auto oldType = func->getLocalType(i);
+        auto newType = Type::getLeastUpperBound(types);
+        assert(newType != Type::none); // in valid wasm there must be a LUB
+
+        // Remove non-nullability, which can't be stored in a local.
+        if (newType.isRef() && !newType.isNullable()) {
+          newType = Type(newType.getHeapType(), Nullable);
+          // Note that the old type must have been nullable as well, as non-
+          // nullable types cannot be locals, which means that we will not have
+          // to do any extra work to handle non-nullability if we update the
+          // type: we are just updating the heap type, and leaving the type
+          // nullable as it was.
+          assert(oldType.isRef() && oldType.isNullable());
+        }
```

Comment:
This extra code could be avoided by including the original type of the local in the LUB calculation. All upper bounds of a nullable type are nullable. If we ever support non-nullable locals, the current code would be better, though.

Reply:
Wouldn't that defeat the optimization? If the local type is `anyref` and the assigned value is `funcref`, the LUB of those two is `anyref`, so we wouldn't change anything (even though we could).

Or do you mean to include "nullability" in the LUB but without the local's HeapType somehow?

Reply:
Oh oops, you're totally right 👍 I was confused :)